lib/semantics/io.cc

```diff
@@ -0,0 +1,619 @@
+  CheckForRequiredSpecifier(SpecifierKind::Advance,
+      hasCharFormat_ || hasLabelFormat_, "an explicit format"s);  // C1221
+  CheckForProhibitedSpecifier(SpecifierKind::Advance, hasInternalUnit_,
+      "UNIT=internal-file"s);  // C1221
+  CheckForRequiredSpecifier(hasAsynchronousYes_, "ASYNCHRONOUS='YES'"s,
+      hasNumberUnit_, "UNIT=number"s);  // C1224
+  CheckForRequiredSpecifier(SpecifierKind::Id, hasAsynchronousYes_,
+      "ASYNCHRONOUS='YES'"s);  // C1225
+  CheckForProhibitedSpecifier(SpecifierKind::Pos, SpecifierKind::Rec);  // C1226
+  // C1227
+  bool fmtOrNml{specifierSet_.test(SpecifierKind::Fmt) ||
+      specifierSet_.test(SpecifierKind::Nml)};
+  CheckForRequiredSpecifier(SpecifierKind::Decimal, fmtOrNml, fmtOrNmlString);
+  CheckForRequiredSpecifier(SpecifierKind::Round, fmtOrNml, fmtOrNmlString);
+}
+
+void IoChecker::SetSpecifier(SpecifierKind specKind) {
+  if (stmt_ == IoStmtKind::None) {
+    // FMT may appear on PRINT statements, which don't have any checks.
+    // [IO]MSG and [IO]STAT parse symbols are shared with non-I/O statements.
+    return;
+  }
+  // C1203, C1207, C1210, C1236, C1239, C1242, C1245
+  if (specifierSet_.test(specKind)) {
+    context_.Say("duplicate %s specifier"_err_en_US,
+        parser::ToUpperCaseLetters(EnumToString(specKind)).data());
+  }
+  specifierSet_.set(specKind);
+}
+
+void IoChecker::CheckStringValue(SpecifierKind specKind,
+    const std::string &value, const parser::CharBlock &source) const {
+  static std::unordered_map<SpecifierKind, const std::vector<std::string>>
+      specValues{
+          {SpecifierKind::Access, {"DIRECT", "SEQUENTIAL", "STREAM"}},
+          {SpecifierKind::Action, {"READ", "READWRITE", "WRITE"}},
+          {SpecifierKind::Advance, {"NO", "YES"}},
+          {SpecifierKind::Asynchronous, {"NO", "YES"}},
+          {SpecifierKind::Blank, {"NULL", "ZERO"}},
+          {SpecifierKind::Decimal, {"COMMA", "POINT"}},
+          {SpecifierKind::Delim, {"APOSTROPHE", "NONE", "QUOTE"}},
+          {SpecifierKind::Encoding, {"DEFAULT", "UTF-8"}},
+          {SpecifierKind::Form, {"FORMATTED", "UNFORMATTED"}},
+          {SpecifierKind::Pad, {"NO", "YES"}},
+          {SpecifierKind::Position, {"APPEND", "ASIS", "REWIND"}},
+          {SpecifierKind::Round,
+              {"COMPATIBLE", "DOWN", "NEAREST", "PROCESSOR_DEFINED", "UP",
+                  "ZERO"}},
+          {SpecifierKind::Sign, {"PLUS", "PROCESSOR_DEFINED", "SUPPRESS"}},
+          {SpecifierKind::Status,
+              // Open values; Close values are different ("DELETE", "KEEP").
+              {"NEW", "OLD", "REPLACE", "SCRATCH", "UNKNOWN"}},
+          {SpecifierKind::Convert, {"BIG_ENDIAN", "LITTLE_ENDIAN", "NATIVE"}},
+          {SpecifierKind::Dispose, {"DELETE", "KEEP"}},
+      };
+  const std::vector<std::string> &valueList{specValues[specKind]};
+  if (std::find(valueList.begin(), valueList.end(),
+          parser::ToUpperCaseLetters(value)) == valueList.end()) {
```

Comment:
Using `std::unordered_map<SpecifierKind, std::set<std::string>>` might be a little simpler.
Then rather than using find it would be:
`if (!specValues.at(specKind).count(parser::ToUpperCaseLetters(value))) {`

Use `.at()` if you want to assert that specKind is one of the kinds in the map.

Reply:
Less is more.  Done.